File: skills/code-architect/references/code-examples/java/Visitor.java
// Visitor — separate algorithms from objects via accept/visit (double dispatch).

interface ShapeVisitor {
    String visitSquare(Square s);
    String visitCircle(Circle c);
}

interface Element {
    String accept(ShapeVisitor v);
}

class Square implements Element {
    final int side;
    Square(int side) { this.side = side; }
    public String accept(ShapeVisitor v) { return v.visitSquare(this); }
}

class Circle implements Element {
    final int radius;
    Circle(int radius) { this.radius = radius; }
    public String accept(ShapeVisitor v) { return v.visitCircle(this); }
}

// One algorithm over the hierarchy, kept outside the element classes.
class AreaVisitor implements ShapeVisitor {
    public String visitSquare(Square s) { return "square area=" + (s.side * s.side); }
    public String visitCircle(Circle c) {
        return "circle area=" + String.format("%.2f", Math.PI * c.radius * c.radius);
    }
}

public class Visitor {
    public static void main(String[] args) {
        Element[] shapes = { new Square(4), new Circle(3) };
        ShapeVisitor area = new AreaVisitor();
        for (Element shape : shapes) {
            System.out.println(shape.accept(area));
        }
    }
}

File: skills/code-architect/references/code-examples/cpp/Visitor.cpp
// Visitor — separate algorithms from objects via accept/visit (double dispatch).
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Circle;
class Square;

// Visitor interface declares a visit per concrete element.
class Visitor {
public:
    virtual ~Visitor() = default;
    virtual void visit(const Circle& c) = 0;
    virtual void visit(const Square& s) = 0;
};

class Element {
public:
    virtual ~Element() = default;
    virtual void accept(Visitor& v) const = 0;
};

class Circle : public Element {
public:
    double radius;
    explicit Circle(double r) : radius(r) {}
    void accept(Visitor& v) const override { v.visit(*this); }
};

class Square : public Element {
public:
    double side;
    explicit Square(double s) : side(s) {}
    void accept(Visitor& v) const override { v.visit(*this); }
};

// Concrete visitor implements one algorithm across all element types.
class AreaVisitor : public Visitor {
public:
    void visit(const Circle& c) override {
        std::cout << "circle area = " << 3.14159 * c.radius * c.radius << "\n";
    }
    void visit(const Square& s) override {
        std::cout << "square area = " << s.side * s.side << "\n";
    }
};

int main() {
    std::vector<std::unique_ptr<Element>> shapes;
    shapes.push_back(std::make_unique<Circle>(2.0));
    shapes.push_back(std::make_unique<Square>(3.0));

    AreaVisitor visitor;
    for (const auto& shape : shapes) shape->accept(visitor);
    return 0;
}
